scipy/datasets/tests/test_data.py

```diff
@@ -0,0 +1,48 @@
+import pooch
+from scipy.datasets._registry import registry
+from scipy.datasets._fetchers import data
+from scipy.datasets import ascent, face, electrocardiogram
+from numpy.testing import assert_equal, assert_almost_equal
+import os
+
+
+def _has_hash(path, expected_hash):
+    """Check if the provided path has the expected hash."""
+    if not os.path.exists(path):
+        return False
+    return pooch.file_hash(path) == expected_hash
+
+
+def test_download_all():
```

Comment:
I noticed an issue here: the other tests are dependent on this one, and hence the tests will fail if run in parallel. E.g., try (after clearing the cache):
```
$ python dev.py test -s datasets --parallel 4
```
To address this, we may want to start using [pytest-dependency](https://pytest-dependency.readthedocs.io/en/stable/about.html). I think we perhaps may have other uses for those in the future (e.g. avoiding a ton of failures for highly parametrized tests).

Alternatively, all these tests can be put in a class with a `download_all()` call in the `setup` method (or whatever the thing is called that is executed once, rather than once per test.

Reply:
Shouldn't we have a `scipy.datasets.download_all()` function by the way?

Reply:
I see your point, but these functions are also supposed to download the datasets individually if there is no cached dataset present. Hence I don't think there is any dependency.

In fact, now that i think more about this, we should have two cases:
1. One where we actually use the cache for testing the datasets. In that case, indeed the dependency will be on `test_download_all` and pooch should utilize the cache.
2. Second is where we manually clear cache and then run tests for each dataset. In that case, each dataset will use pooch to download the files.

This way, we'll be testing all possible cases for dataset file retrieval.

`python dev.py test -s datasets --parallel 4` works just fine for me, but I'm curious if this can cause some sort of corruption (since more than one function may be downloading the file and writing/overwriting the contents at the same time. I'm guessing here).

> scipy.datasets.download_all()

Yeah, honestly, I see the use of that only once in the lifetime of scipy installation for a user, that too when someone wants to make sure that their SciPy installation always works without the internet connection. Seems like a very special case, but it wouldn't really hurt to add this utility function. Can be discussed, should be an easy addition with no maintenance anyway.

Reply:
We discussed the race condition in person just now, and also that it's desirable not to download the same dataset twice. I'll wait for your next update here.

Reply:
Done, except the second case is not being tested right now, will add that later when we have a utility function to clear all cache.